fix mysql datetime type according to the column size

diqksrk opened the pull request:
## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
<!--  Maintainers only: Mandatory Labels to add to a PR

At least one of the labels must be added to the PR before it's merged. If no label is provided the workflow will fail and you will not be able to merge the PR. After the label is added it re-runs the `Pull Request Labels / label (pull_request)` and gives a green check. 

`skipReleaseNotes`   - Don't show up on the Draft Release Notes page
`notableChanges`     - Any notable changes
`TypeEnhancement`    - New features
`TypeTest`           - New Test features
`TypeBug`            - bug fixes
`breakingChanges`    - any breaking changes
`APIBreakingChanges` - any API breaking changes
`sdou`               - Security, Driver and Other Updates -dependabot PR's
`newContributors`    - New Contributors 

-->


## Description
Fixes #6069

In the case of MySQL, the `datetime` and `timestamp` types are affected by the column size. The type, such as datetime, datetime(1), or datetime(2), is determined according to the column size, not decimal_digits.

Therefore, I am submitting a PR that determines the type of datetime based on the column size as follows

Here are some additional materials.
https://forums.mysql.com/read.php?3,506409,506409
![image](https://github.com/liquibase/liquibase/assets/31757314/94eb33bc-d198-40b3-ab36-f0340f318bde)

https://dev.mysql.com/doc/refman/8.4/en/storage-requirements.html
![image](https://github.com/liquibase/liquibase/assets/31757314/4d141e64-5ced-40a1-9586-f0abd669064c)


<!--
A clear and concise description of the change being made.  

- Introduce what was/will be done in the title
  - Titles show in release notes and search results, so make them useful
  - Use verbs at the beginning of the title, such as "fix", "implement", "improve", "update", and "add" 
  - Be specific about what was fixed or changed
  - Good Example: `Fix the --should-snapshot-data CLI parameter to be preserved when the --data-output-directory property is not specified in the command.`
  - Bad Example: `Fixed --should-snapshot-data`  
- If there is an existing issue this addresses, include "Fixes #XXXX" to auto-link the issue to this PR
- If there is NOT an existing issue, consider creating one.
  - In general, issues describe wanted change from an end-user perspective and PRs describe the technical change.
  - If this change is very small and not worth splitting off an issue, include `Steps To Reproduce`, `Expected Behavior`, and `Actual Behavior` sections in this PR as you would have in the issue.
- Describe what users need and how the fix will affect them
- Describe how the code change addresses the problem
- Ensure private information is redacted.
-->

## Things to be aware of

<!--
- Describe the technical choices you made
- Describe impacts on the codebase
-->

## Things to worry about

<!--
- List any questions or concerns you have with the change
- List unknowns you have 
-->

## Additional Context

<!--
Add any other context about the problem here.
-->

tati-qalified commented on liquibase-standard/src/main/java/liquibase/database/core/MySQLDatabase.java:
```diff
@@ -273,6 +271,25 @@ public int getDatabasePatchVersion() throws DatabaseException {
 
     }
 
+    public Integer getTimeStampFSPWith(int columnSize) {
```
I think this could be improved in terms of efficiency. I propose adding an "if" statement and then just returning columnSize%10.

It would be something along the lines of:
```
public Integer getTimeStampFSPWith(int columnSize) {
    if (columnSize > 20 && columnSize < 27) {
        return columnSize % 10
    }
    return 0
}
```

I'm unaware of the lower and upper limits on the columnSize, so it might not need both verifications.

diqksrk replied:
Thank you very much for your advice.
I tested the method and confirmed that it works correctly.
After changing to this method, I proceeded to commit again.

MalloD12 commented on liquibase-standard/src/test/groovy/liquibase/snapshot/jvm/ColumnSnapshotGeneratorTest.groovy:
```diff
@@ -49,6 +50,22 @@ class ColumnSnapshotGeneratorTest extends Specification {
         dataType.getTypeName() == "varchar"
     }
 
+    def "ReadDataType specifies column size for mysql datetime data fsp"() {
```
I think we can customize a bit this test and make it lightly generic by parameterizing `database`, `columnSize`, and `fps`. Then we can list different scenarios with different values, and test different limits and DBs. At this point, this change is applicable for MySQL and also for MariaDB (which extends from MySQL) and the test result would be the same for either of them, but if there is any future change or any method override it would be nice to be testing the two DBs separately. If you need some help with Spock details we can assist you with it.

MalloD12 replied:
This would be also nice to do, but it's not mandatory. I can do it if you don't mind.

Thanks,
Daniel.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/snapshot/jvm/ColumnSnapshotGenerator.java:
```diff
@@ -501,6 +501,10 @@
             }
 
+            if (database instanceof MySQLDatabase) {
+                type.setColumnSize(((MySQLDatabase)database).getTimeStampFSPWith(columnSize));
```
I think it would be nice to add a null check on this `if` condition.

MalloD12 commented:
@rberezen - @tati-qalified: I think besides the comment I added about the existent test of this PR we are good with the test coverage of this change. I was thinking additionally that we can add an integration test to generate a snapshot and then validate type is what we expect, but maybe it's not worth. What do you think?

@diqksrk: if i'm not wrong this is only applicable for `datetime`. But what about time and timestamp or any other type that we can specify a `FSP`? Should we not consider making a type distinction in the given method (`getTimeStampFSPWith`)?
And one more thing, considering we are returning the FSP I was wondering whether we should change this method name to something like `getFSPFromTimeType`?

Thank you, guys!

rberezen commented:
> @rberezen - @tati-qalified: I think besides the comment I added about the existent test of this PR we are good with the test coverage of this change. I was thinking additionally that we can add an integration test to generate a snapshot and then validate type is what we expect, but maybe it's not worth. What do you think?
> 
> @diqksrk: if i'm not wrong this is only applicable for `datetime`. But what about time and timestamp or any other type that we can specify a `FSP`? Should we not consider making a type distinction in the given method (`getTimeStampFSPWith`)? And one more thing, considering we are returning the FSP I was wondering whether we should change this method name to something like `getFSPFromTimeType`?
> 
> Thank you, guys!

I think we do not need extra tests here.

diqksrk commented:
> @rberezen - @tati-qalified: I think besides the comment I added about the existent test of this PR we are good with the test coverage of this change. I was thinking additionally that we can add an integration test to generate a snapshot and then validate type is what we expect, but maybe it's not worth. What do you think?
> 
> @diqksrk: if i'm not wrong this is only applicable for `datetime`. But what about time and timestamp or any other type that we can specify a `FSP`? Should we not consider making a type distinction in the given method (`getTimeStampFSPWith`)? And one more thing, considering we are returning the FSP I was wondering whether we should change this method name to something like `getFSPFromTimeType`?
> 
> Thank you, guys!

First, I apologize for the delayed response.

In MySQL, the date types that can have an FSP value are TIME, DATETIME, and TIMESTAMP. This can be confirmed in the following document: [MySQL Documentation on Fractional Seconds](https://dev.mysql.com/doc/refman/8.4/en/fractional-seconds.html).

Among these, the types mapped to java.sql.Types.TIMESTAMP = 93 are DATETIME and TIMESTAMP.

As you mentioned, the data type corresponding to java.sql.Types.TIME = 92 is TIME. Both types use FSP to calculate the Column Size.

The way to calculate column size is divided into:

1. TIME
2. DATETIME and TIMESTAMP

Currently, I have implemented the calculation method for the second part. As you suggested, I will modify the method to accept a variable indicating whether it is TIME or DATETIME/TIMESTAMP, and then return the column size accordingly for each type.

I will also investigate and correct any error cases related to the first part 'TIME', and commit the changes.

Thank you always for your kind guidance!

diqksrk commented:
@MalloD12 
As you suggested, I have modified the method to accept jdbctype as a variable.

I will investigate if there are any error cases related to Time, and if any are found, I will register a new issue and inform you.

Thank you.

MalloD12 commented on liquibase-standard/src/main/java/liquibase/snapshot/jvm/ColumnSnapshotGenerator.java:
```diff
@@ -501,6 +501,10 @@ protected DataType readDataType(CachedRow columnMetadataResultSet, Column column
             }
 
+            if (database instanceof MySQLDatabase) {
+                type.setColumnSize(((MySQLDatabase) database).getFSPFromTimeType(columnSize, Types.TIMESTAMP));
```
We need to add a null check for `columnSize` here. That's being reported as an alert from our CodeQL scan. Could you please do that?

Thanks,
Daniel.

diqksrk commented:
@MalloD12 
I have added a null check for columnSize. Additionally, if it is null, it will be set to the default value.

For example, in the case of DATETIME, it can be DATETIME(0), DATETIME(1), DATETIME(2), and so on up to DATETIME(6). Here, it will be set to the default value of 0

MalloD12 reviewed (approved):
**Approved**.

Thanks, @diqksrk for this PR and for addressing our review comments. Code changes look good to me. I made a little change to the test you provided here and make it parameterized.

Thanks,
Daniel.

tati-qalified reviewed (approved):
Verified that this works correctly for all 4 changelog formats.
If the existing table has a column of type `datetime`, with no precision specified, the generated changelog will show `datetime(0)`, which is equivalent.
